### DISCUSSION TITLE
Convert OME-tif to dzi

### DISCUSSION THREAD

**Discussion by miriam-groeneveld** [libvips/libvips#3911] [Q&A]
I have ome-tif files (with colourspace grey16) and need to convert these to either tiff or dzi. I'm not sure what the best way would be, I've tried several methods (tiffsave, write_to_file, dzsave) with several compressions (jp2k), but always end up with black images. I'm hoping someone can point me in the right direction.

I can't share the image, but these are the tif tags:

File Size                       : 2.1 MB
Zone Identifier                 : Exists
File Modification Date/Time     : 2024:03:25 10:11:27+01:00
File Access Date/Time           : 2024:03:26 08:57:43+01:00
File Creation Date/Time         : 2024:03:25 10:11:27+01:00
File Permissions                : -rw-rw-rw-
File Type                       : TIFF
File Type Extension             : tif
MIME Type                       : image/tiff
Exif Byte Order                 : Little-endian (Intel, II)
Image Width                     : 1016
Image Height                    : 980
Bits Per Sample                 : 16
Compression                     : Uncompressed
Photometric Interpretation      : BlackIsZero
Image Description               : <OME xmlns="http://www.openmicroscopy.org/Schemas/OME/2016-06" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.openmicroscopy.org/Schemas/OME/2016-06 http://www.openmicroscopy.org/Schemas/OME/2016-06/ome.xsd" UUID="redacted" Creator="Scientific Volume Imaging B.V., The Huygens System, V23.10.0p3">.  <ExperimenterGroup Name="Study 26" ID="ExperimenterGroup:26"/>.  <Instrument ID="Instrument:0">.    <Objective ID="Objective:0" Immersion="Oil" LensNA="1.25" NominalMagnification="10.0"/>.  </Instrument>.  <Image ID="Image:0" Name="image_25_BF_z0.ome.tiff">.    <ExperimenterGroupRef ID="ExperimenterGroup:26"/>.    <ObjectiveSettings ID="Objective:0" Medium="Oil" RefractiveIndex="1.515"/>.    <Pixels ID="Pixels:0" DimensionOrder="XYZCT" Type="uint16" SignificantBits="16" Interleaved="false" BigEndian="false" SizeX="1016" SizeY="980" SizeZ="1" SizeC="1" SizeT="1" PhysicalSizeX="0.102"  PhysicalSizeY="0.102"  PhysicalSizeZ="1.0"  TimeIncrement="1.0" TimeIncrementUnit="s">.      <Channel ID="Channel:0:0" Name="BF" SamplesPerPixel="1" PinholeSize="5.6"  AcquisitionMode="WideField" ExcitationWavelength="483.0" ExcitationWavelengthUnit="nm" EmissionWavelength="525.0" EmissionWavelengthUnit="nm">.        <LightPath/>.      </Channel>.      <MetadataOnly/>.      <Plane TheZ="1" TheT="0" TheC="0"/>.    </Pixels>.    <AnnotationRef ID="Annotation:CustomAttributes:SVI:Image:0"/>.    <AnnotationRef ID="Annotation:26"/>.  </Image>.  <StructuredAnnotations>.    <XMLAnnotation ID="Annotation:CustomAttributes:SVI:Image:0">.      <Value>.        <xsi:Geometry>.          <xsi:SamplingSizes DeltaT="1.000000" DeltaTState="Default" DeltaX="0.102000" DeltaXState="Reported" DeltaY="0.102000" DeltaYState="Reported" DeltaZ="1.000000" DeltaZState="Reported"/>.          <xsi:Offsets OffsetT="0.000000" OffsetTState="Default" OffsetX="0.000000" OffsetXState="Default" OffsetY="0.000000" OffsetYState="Default" OffsetZ="0.000000" OffsetZState="Default"/>.          <xsi:Interface InterFacePrimary="0.000000" InterFacePrimaryState="Default" InterFaceSecondary="1000000.000000" InterFaceSecondaryState="Default"/>.        </xsi:Geometry>.        <xsi:ChannelData ChannelID="Channel:0" ExPhotonCnt="1" ExPhotonCntState="Estimated" ExcitationBeamFill="2.000000" ExcitationBeamFillState="Default" LambdaEm="525.000000" LambdaEmState="Reported" LambdaEx="483.000000" LambdaExState="Estimated" MicroscopeSpec="Widefield" MicroscopeSpecState="Estimated" ObjectiveQuality="80" ObjectiveQualityState="Default" PinholeSpacing="2.530000" PinholeSpacingState="Default" RefrIndexLensMedium="1.515000" RefrIndexLensMediumState="Estimated" RefrIndexMedium="1.515000" RefrIndexMediumState="Estimated">.          <xsi:MicrParamData DetectorMagnif="10.000000" DetectorMagnifState="Default" IllPinholeRadius="0.000000" IllPinholeRadiusState="Default" ImagingDirection="upward" ImagingDirectionState="Default" Mode="fluorescence" ModeState="Estimated" NumAperture="1.250000" NumApertureState="Default" PinholeRadius="280.000000" PinholeRadiusState="Default"/>.          <xsi:DetectorData Baseline="0.000000" BaselineState="Default" LineAvgCnt="1.000000" LineAvgCntState="Default" NoiseGain="1.000000" NoiseGainState="Default" PPU="1.000000" PPUState="Default"/>.          <xsi:ScatterData Blurring="0.000000" BlurringState="Default" FreePath="1000.000000" FreePathState="Default" Model="none" ModelState="Default" RelContrib="0.000000" RelContribState="Default"/>.        </xsi:ChannelData>.        <xsi:ChannelData ChannelID="Channel:1" ExPhotonCnt="1" ExPhotonCntState="Estimated" ExcitationBeamFill="2.000000" ExcitationBeamFillState="Default" LambdaEm="685.000000" LambdaEmState="Reported" LambdaEx="630.200012" LambdaExState="Estimated" MicroscopeSpec="Widefield" MicroscopeSpecState="Estimated" ObjectiveQuality="80" ObjectiveQualityState="Default" PinholeSpacing="2.530000" PinholeSpacingState="Default" RefrIndexLensMedium="1.515000" RefrIndexLensMediumState="Estimated" RefrIndexMedium="1.515000" RefrIndexMediumState="Estimated">.          <xsi:MicrParamData DetectorMagnif="10.000000" DetectorMagnifState="Default" IllPinholeRadius="0.000000" IllPinholeRadiusState="Default" ImagingDirection="upward" ImagingDirectionState="Default" Mode="fluorescence" ModeState="Estimated" NumAperture="1.250000" NumApertureState="Default" PinholeRadius="280.000000" PinholeRadiusState="Default"/>.          <xsi:DetectorData Baseline="0.000000" BaselineState="Default" LineAvgCnt="1.000000" LineAvgCntState="Default" NoiseGain="1.000000" NoiseGainState="Default" PPU="1.000000" PPUState="Default"/>.          <xsi:ScatterData Blurring="0.000000" BlurringState="Default" FreePath="1000.000000" FreePathState="Default" Model="none" ModelState="Default" RelContrib="0.000000" RelContribState="Default"/>.        </xsi:ChannelData>.        <xsi:PDimConfig Size="1" Type="generic">.        </xsi:PDimConfig>.        <xsi:FeatureHistory FeatureCnt="3" HuygensVersion="23.10.0p3" OmeWriterVersion="1.1p0">.          <xsi:FeatureDescription String="none;"/>.          <xsi:FeatureDescription String="base;"/>.          <xsi:FeatureDescription String="Parameter states and STED, SPIM and pdim parameters;"/>.        </xsi:FeatureHistory>.      </Value>.    </XMLAnnotation>.    <MapAnnotation ID="Annotation:26" Namespace="BioStudies">.      <Value>.        <M K="study">26</M>.      </Value>.    </MapAnnotation>.  </StructuredAnnotations>.</OME>
Samples Per Pixel               : 1
X Resolution                    : 0.102
Y Resolution                    : 0.102
Resolution Unit                 : inches
Software                        : tifffile.py
Tile Width                      : 128
Tile Length                     : 128
Tile Offsets                    : (Binary data 476 bytes, use -b option to extract)
Tile Byte Counts                : (Binary data 383 bytes, use -b option to extract)
Image Size                      : 1016x980
Megapixels                      : 0.996

**Comment by jcupitt** [comment]
Hi @miriam-groeneveld,

I think `dzsave` should work. What libvips version are you using? Do you have a sample image you can share? What platform are you on?

**Reply by miriam-groeneveld** [reply]
I'm on Windows 11, vips 8.14.1. I can't share the image, but updated the initial description with the tif info.

**Reply by miriam-groeneveld** [reply]
One warning I do get when using dzsave:

> (vips:9816): VIPS-WARNING **: 12:11:26.955: Invalid TIFF directory; tags are not sorted in ascending order
(vips:9816): VIPS-WARNING **: 12:11:26.961: Invalid TIFF directory; tags are not sorted in ascending order
(vips:9816): VIPS-WARNING **: 12:11:26.967: Invalid TIFF directory; tags are not sorted in ascending order
(vips:9816): VIPS-WARNING **: 12:11:26.973: Invalid TIFF directory; tags are not sorted in ascending order
(vips:9816): VIPS-WARNING **: 12:11:26.973: Invalid TIFF directory; tags are not sorted in ascending order

**Reply by jcupitt** [reply]
I tried here:

```
$ vips colourspace nina.jpg x.tif grey16
$ tiffinfo x.tif
=== TIFF directory 0 ===
TIFF Directory at offset 0x2e83008 (48771080)
  Image Width: 6048 Image Length: 4032
  Resolution: 300, 300 pixels/inch
  Bits/Sample: 16
  Sample Format: unsigned integer
  Compression Scheme: None
  Photometric Interpretation: min-is-black
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 1
  Rows/Strip: 128
  Planar Configuration: single image plane
$ vips dzsave x.tif x
```

Then `x_files/8/0_0.jpeg` is:

![0_0](https://github.com/libvips/libvips/assets/580843/676b2a42-f511-4783-a4d4-9b938303b59e)

So it seems to work. What command did you run?

**Reply by jcupitt** [reply]
> (vips:9816): VIPS-WARNING **: 12:11:26.955: Invalid TIFF directory; tags are not sorted in ascending order

You have an invalid TIFF file, I guess there's a bug in `tifffile.py`, though it probably doesn't matter.

**Reply by jcupitt** [reply]
Perhaps your TIFF file is not using 0 - 65535 for black to white? Do you know the numeric range it has (ie. min pixel to max pixel value)?

**Reply by miriam-groeneveld** [reply]
I use the `vips dzsave` command. I have the same issue (without the warning) when using one of the sample files, like this one: https://downloads.openmicroscopy.org/images/OME-TIFF/2016-06/BBBC017/single-file/

**Reply by jcupitt** [reply]
You can show the numeric range with eg.:

```
$ vips colourspace nina.jpg x.tif grey16
$ vips stats x.tif .csv
0	65535	743040180717	34138482502869176	30470.528952777582	21713.933104220141	1525	591	48	0
0	65535	743040180717	34138482502869176	30470.528952777582	21713.933104220141	1525	591	48	0
```

Columns are minimum, maximum, sum, sum of squares, mean, standard deviation, x coordinate of minimum, y coordinate of minimum, x coordinate of maximum, y coordinate of maximum. The first row is for all bands, subsequent rows are for the image bands in order.

**Reply by jcupitt** [reply]
Ah your file is not using the correct numeric range. I tried in vipsdisp and I see:

![image](https://github.com/libvips/libvips/assets/580843/4691b2b9-9282-4f3c-913a-b095ded1a0b5)

It's a 16-bit image, but it's mostly using 0 - 255 for black to white (you can see the pixel values in the status bar at the bottom of the window -- each column is a page or channel). When libvips converts to 8 bit for dzsave, it looks almost black.

You'll need to scale the numeric range to fit 0 - 65535. I would use something like python to preprocess the image, if you have that available. Perhaps:

```python
import pyvips

image = pyvips.Image.new_from_file("NIRHTa+001.ome.tiff")
image = image * 256
image.dzsave("x")
```

**Reply by miriam-groeneveld** [reply]
Thanks John, now I do have a better image with this code snippet:

```
im = pyvips.Image.new_from_file(resource)
im2 = im * 256

im2.write_to_file(
  output_dir / "output.tif",
  tile=True,
  pyramid=True,
)
im2.dzsave(output_dir / "output.dzi")
```
It does look like only a single level gets saved and converted to dzi, can that be right?

**Reply by jcupitt** [reply]
OME TIFF uses the "pages" dimension of TIFF files to store the image channels (also called bands), so eg. R, G, B would be in pages 0, 1 and 2. That sample image you linked above has almost 7,000 pages (!!). DZI uses JPEG for tiles (by default), so (usually) only three channels. You need to think about how many channels your OME TIFF has and how you want to map those to the available RGB in DZI. 

`pyvips.Image.new_from_file` only takes the first (zeroth) page by default, but you can use the `page` and `n` parameters to pick a start page and a number of pages.

You could take the first three pages from your OME TIFF like this:

```python
# this will load the three pages as a tall, thin grey16 image, like
# (sorry) three sheets of toilet paper joined together 
image = pyvips.Image.new_from_file(resource, page=0, n=3)

# the height of each page in the tall, thin image
page_height = image.get("page-height")

# chop into pages
pages = [image.crop(0, y, image.width, page_height)
         for y in range(0, image.height, page_height)]

# join pages band-wise to make an interleaved image
image = pages[0].bandjoin(pages[1:])

# tag the image as RGB
image = image.copy(interpretation="srgb")

# now save and you'll get an RGB DZI ... you might need to adjust the numeric range
# of the channels
image.dzsave("output.dzi")
```

**Comment by jcupitt** [comment]
There's a windows build of vipsdisp here fwiw: https://github.com/jcupitt/vipsdisp/releases/tag/v3.0.4

**Reply by miriam-groeneveld** [reply]
I can open the file in vispdisp, but it's also black in that application:
![image](https://github.com/libvips/libvips/assets/51329283/de7c1d19-5997-4a25-9db9-9e038f6a4c15)

**Reply by miriam-groeneveld** [reply]
ah, nvm, found the controls :).

**Reply by jcupitt** [reply]
The controls have things like falsecolour as well, which can be handy, you probably saw.

If you select "Toilet roll" (what a terrible name) you can scroll up and down to see all the pages.

Use shift + cursor up/down to scroll a page at a time.

**Reply by miriam-groeneveld** [reply]
Agree on the terrible naming 👍. Thanks for the help so far, this is a great step forward already! Did you notice my question about the dzi creation of single levels (https://github.com/libvips/libvips/discussions/3911#discussioncomment-8914584)?

**Reply by jcupitt** [reply]
Yes, I replied ^^^ maybe refresh your browser?

**Reply by miriam-groeneveld** [reply]
Ah, I'm so sorry, you are right. Thank you for the explanation, you've been a great help!
